aws-lc-rs/src/ec.rs

```diff
@@ -362,40 +391,83 @@ pub(crate) unsafe fn unmarshal_der_to_private_key(
 }
 
 pub(crate) unsafe fn marshal_public_key_to_buffer(
-    buffer: &mut [u8; PUBLIC_KEY_MAX_LEN],
-    evp_pkey: &ConstPointer<EVP_PKEY>,
+    buffer: &mut [u8],
+    evp_pkey: &LcPtr<EVP_PKEY>,
+    compressed: bool,
 ) -> Result<usize, Unspecified> {
-    let ec_key = ConstPointer::new(EVP_PKEY_get0_EC_KEY(**evp_pkey))?;
-    marshal_ec_public_key_to_buffer(buffer, &ec_key)
+    let ec_key = ConstPointer::new(EVP_PKEY_get0_EC_KEY(evp_pkey.as_const_ptr()))?;
+    marshal_ec_public_key_to_buffer(buffer, &ec_key, compressed)
 }
 
 pub(crate) unsafe fn marshal_ec_public_key_to_buffer(
-    buffer: &mut [u8; PUBLIC_KEY_MAX_LEN],
+    buffer: &mut [u8],
     ec_key: &ConstPointer<EC_KEY>,
+    compressed: bool,
 ) -> Result<usize, Unspecified> {
     let ec_group = ConstPointer::new(EC_KEY_get0_group(**ec_key))?;
 
     let ec_point = ConstPointer::new(EC_KEY_get0_public_key(**ec_key))?;
 
-    let out_len = ec_point_to_bytes(&ec_group, &ec_point, buffer)?;
+    let point_conversion_form = if compressed {
+        point_conversion_form_t::POINT_CONVERSION_COMPRESSED
+    } else {
+        point_conversion_form_t::POINT_CONVERSION_UNCOMPRESSED
+    };
+
+    let out_len = ec_point_to_bytes(&ec_group, &ec_point, buffer, point_conversion_form)?;
     Ok(out_len)
 }
 
-pub(crate) fn marshal_public_key(
-    evp_pkey: &ConstPointer<EVP_PKEY>,
+pub(crate) fn public_key_from_evp_pkey(
+    evp_pkey: &LcPtr<EVP_PKEY>,
     algorithm: &'static EcdsaSigningAlgorithm,
 ) -> Result<PublicKey, Unspecified> {
     let mut pub_key_bytes = [0u8; PUBLIC_KEY_MAX_LEN];
     unsafe {
-        let key_len = marshal_public_key_to_buffer(&mut pub_key_bytes, evp_pkey)?;
+        let key_len = marshal_public_key_to_buffer(&mut pub_key_bytes, evp_pkey, false)?;
 
         Ok(PublicKey {
+            evp_pkey: evp_pkey.clone(),
             algorithm,
             octets: pub_key_bytes[0..key_len].into(),
         })
     }
 }
 
+pub(crate) fn try_parse_public_key_bytes(
+    key_bytes: &[u8],
+    expected_curve_nid: i32,
+) -> Result<LcPtr<EVP_PKEY>, Unspecified> {
+    if let Ok(key) = try_parse_subject_public_key_info_bytes(key_bytes) {
+        return Ok(validate_evp_key(&key.as_const(), expected_curve_nid).map(|()| key)?);
+    }
+    // try_parse_public_key_raw_bytes calls `validate_evp_key` already
+    try_parse_public_key_raw_bytes(key_bytes, expected_curve_nid)
+}
```

Comment:
The following alternative implementation might be more idiomatic, but it seems less readable to me:
```
pub(crate) fn try_parse_public_key_bytes(
    key_bytes: &[u8],
    expected_curve_nid: i32,
) -> Result<LcPtr<EVP_PKEY>, Unspecified> {
    try_parse_subject_public_key_info_bytes(key_bytes)
        .and_then(|key| {
            validate_evp_key(&key.as_const(), expected_curve_nid)
                .map(|()| key)
                .map_err(|_| Unspecified)
        })
        .or(try_parse_public_key_raw_bytes(
            key_bytes,
            expected_curve_nid,
        ))
}
```